### PR TITLE
Make CoreController actions CaaSes

### PR DESCRIPTION
**Pull request by greg0ire**
I am targeting this branch, because this is BC.

## Changelog

```markdown
### Changed
- `Controller\CoreController` is now deprecated in favor of `Action\{Search,Dasbhoard}Action`
```

## Todo
- [x] unit test search action
- [x] add `NEXT_MAJOR` comments
- [x] move to `Action` namespace
- [x] fix the build
- [x] rebase on #5122
- [x] test this on an actual project

## Subject

Using invokable CaaSes vs one big controller achieves several things:
1. it discourages piling on more actions in controllers with vague names (SRP)
2. it leads to easier to unit test actions and visible deps (DIP)
3. it will allow us to make more services private

### CODE REVIEW

File: src/Controller/SearchAction.php [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+     * The search action first render an empty page, if the query is set, then the template generates
+     * some ajax request to retrieve results for each admin. The Ajax query returns a JSON response.
+     *
+     * @throws \RuntimeException
+     *
+     * @return JsonResponse|Response
+     */
+    public function __invoke(Request $request)
+    {
+        if ($request->get('admin') && $request->isXmlHttpRequest()) {
```
Comment (kunicmarko20) [inline src/Controller/SearchAction.php]: What about inversing if and putting the default return instead of this.

File: src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php [hunk 2]
```diff
@@ -378,7 +378,7 @@ public function fixTemplates(
 
             // set template for simple pager if it is not already overwritten
             if ('setPagerType' === $method[0]
-                && $method[1][0] === Pager::TYPE_SIMPLE
+                && Pager::TYPE_SIMPLE === $method[1][0]
```
Comment (kunicmarko20) [inline src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: Is this in scope of this PR?

File: src/Controller/DashboardAction.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+namespace Sonata\AdminBundle\Controller;
+
+use Sonata\AdminBundle\Admin\BreadcrumbsBuilderInterface;
+use Sonata\AdminBundle\Admin\Pool;
+use Sonata\AdminBundle\Templating\TemplateRegistryInterface;
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\Request;
+
+final class DashboardAction extends Controller
```
Comment (core23) [inline src/Controller/DashboardAction.php]: Why do you extend the symfony `Controller`?

File: src/Controller/DashboardAction.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\AdminBundle\Controller;
```
Comment (core23) [inline src/Controller/DashboardAction.php]: Not a blocker, but I would prefer the `..\Action` namespace for actions.

Comment (kunicmarko20) [reply src/Controller/DashboardAction.php]: Then maybe also change the name of service file to actions?

File: src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php (hunk 2 above)
Comment (greg0ire) [reply src/DependencyInjection/Compiler/AddDependencyCallsCompilerPass.php]: It's cs

File: src/Controller/DashboardAction.php (hunk 3 above)
Comment (greg0ire) [reply src/Controller/DashboardAction.php]: For one thing and one thing only: the `render` method. I looked at it, it is fairly complicated because it looks for the `templating` component, and for `twig`. Not sure if I could safely get rid of it.

**Comment by greg0ire** [comment]
@core23 why did you relabel it to minor?

I'm going to make another PR for cs.

**Comment by greg0ire** [comment]
> @core23 why did you relabel it to minor?

Ah I know, that's because of the deprecations, forgot that rule :+1:

**Comment by core23** [comment]
> Ah I know, that's because of the deprecations, forgot that rule 👍

That's one reason. You also introduce new stuff (the action classes), so it's a minor.

**Comment by greg0ire** [comment]
I introduce new classes, but I introduce no new features I think :P

File: src/Controller/SearchAction.php (hunk 1 above)
Comment (greg0ire) [reply src/Controller/SearchAction.php]: I just did this, good call :)

**Comment by greg0ire** [comment]
Made Codacy happy :)

File: src/DependencyInjection/SonataAdminExtension.php [hunk 5]
```diff
@@ -141,6 +141,7 @@ public function load(array $configs, ContainerBuilder $container)
         $loader->load('block.xml');
         $loader->load('menu.xml');
         $loader->load('commands.xml');
+        $loader->load('actions.xml');
```
Comment (core23) [inline src/DependencyInjection/SonataAdminExtension.php]: action.xml

Comment (greg0ire) [reply src/DependencyInjection/SonataAdminExtension.php]: Why use singular? We use Commands and services

Comment (core23) [reply src/DependencyInjection/SonataAdminExtension.php]: What about block and menu?

Comment (greg0ire) [reply src/DependencyInjection/SonataAdminExtension.php]: Maybe those are faulty, or maybe there is an unspoken rule about when to use singular and when to use plural, all I know is that in Symfony, you have `services.yaml`. Either way, fixing this is out of scope of this PR.

Comment (core23) [reply src/DependencyInjection/SonataAdminExtension.php]: Symfony is refering to a `services.yaml`:
https://symfony.com/doc/current/service_container.html#service-container-services-load-example

So we should use plurals 👍

**Comment by greg0ire** [comment]
Finally managed to test it on a very old project, it doesn't work :P

**Comment by greg0ire** [comment]
Ready for review :)

**Comment by greg0ire** [comment]
greg0ire requested a review (and also, a merge would be nice) from sonata-project/contributors 2 minutes ago

**Comment by covex-nn** [comment]
Thank you, @greg0ire =)

**Comment by greg0ire** [comment]
@greg0ire is happy :)
